kratos/processes/fast_transfer_between_model_parts_process.cpp

```diff
@@ -53,91 +61,282 @@ void FastTransferBetweenModelPartsProcess::Execute()
+    const int num_nodes = static_cast<int>(mrOriginModelPart.Nodes().size());
+    const SizeType total_num_elements = root_model_part.Elements().size();
+    const int num_elements = static_cast<int>(mrOriginModelPart.Elements().size());
+    const SizeType total_num_conditions = root_model_part.Conditions().size();
+    const int num_conditions = static_cast<int>(mrOriginModelPart.Conditions().size());
 
-                if (num_elements != 0 && (mEntity == EntityTransfered::ALL || mEntity == EntityTransfered::ELEMENTS || mEntity == EntityTransfered::NODESANDELEMENTS))
-                    for( auto& element_buffer : elements_buffer)
-                        mrDestinationModelPart.AddElements(element_buffer.begin(),element_buffer.end());
+    #pragma omp parallel
+    {
+        const int id = OpenMPUtils::ThisThread();
+
+        if (num_nodes != 0 && (mEntity == EntityTransfered::ALL || mEntity == EntityTransfered::NODES || mEntity == EntityTransfered::NODESANDELEMENTS || mEntity == EntityTransfered::NODESANDCONDITIONS)) {
+            #pragma omp for
+            for(int i = 0; i < num_nodes; ++i) {
+                auto it_node = mrOriginModelPart.NodesBegin() + i;
+                NodeType::Pointer p_new_node = it_node->Clone();
+                p_new_node->SetId(total_num_nodes + i + 1);
+                (nodes_buffer[id]).insert((nodes_buffer[id]).begin(), p_new_node);
+            }
+        }
+
+        if (num_elements != 0 && (mEntity == EntityTransfered::ALL || mEntity == EntityTransfered::ELEMENTS || mEntity == EntityTransfered::NODESANDELEMENTS)) {
+            #pragma omp for
+            for(int i = 0; i < num_elements; ++i) {
+                auto it_elem = mrOriginModelPart.ElementsBegin() + i;
+                Element::Pointer p_new_elem = it_elem->Clone(total_num_elements + i + 1, it_elem->GetGeometry());
+                (elements_buffer[id]).insert((elements_buffer[id]).begin(), p_new_elem);
+            }
+        }
 
-                if (num_conditions != 0 && (mEntity == EntityTransfered::ALL || mEntity == EntityTransfered::CONDITIONS || mEntity == EntityTransfered::NODESANDCONDITIONS))
-                    for( auto& condition_buffer : conditions_buffer)
-                        mrDestinationModelPart.AddConditions(condition_buffer.begin(),condition_buffer.end());
+        if (num_conditions != 0 && (mEntity == EntityTransfered::ALL || mEntity == EntityTransfered::CONDITIONS || mEntity == EntityTransfered::NODESANDCONDITIONS)) {
+            #pragma omp for
+            for(int i = 0; i < num_conditions; ++i) {
+                auto it_cond = mrOriginModelPart.ConditionsBegin() + i;
+                Condition::Pointer p_new_cond = it_cond->Clone(total_num_conditions + i + 1, it_cond->GetGeometry());
+                (conditions_buffer[id]).insert((conditions_buffer[id]).begin(), p_new_cond);
             }
         }
+
+        // We add to the model part
+        #pragma omp single
```

Comment:
You may want to declare 3 different single regions so they can be executed in parallel if there are enough cores

Reply:
Ok, wi will solve it